[TableGen] Added submulticlass typechecking to template arg values.

jofrn opened the pull request:
Some typechecking was missing when parsing a submulticlass reference. This adds it in and fixes any mistyped codes in .td files.

jofrn committed:
[TableGen] Added submulticlass typechecking to template arg values.

Some typechecking was missing when parsing a submulticlass reference.
This adds it in and fixes any mistyped codes in .td files.

jofrn commented:
This resolves https://github.com/llvm/llvm-project/issues/84910.

jayfoad commented on llvm/lib/Target/AMDGPU/VOPInstructions.td:
```diff
@@ -37,7 +37,7 @@ defvar VOPDX_Max_Index = 12;
 class VOPD_Component<bits<5> OpIn, string vOPDName> {
   Instruction BaseVOP = !cast<Instruction>(NAME);
-  string VOPDName = "v_dual_" # !substr(vOPDName, 2);
+  string VOPDName = "v_dual_" # !if(!le(!size(vOPDName), 2), "", !substr(vOPDName, 2));
```
How is this change related to template arg type checking?

jofrn replied:
A def within a multiclass inherits from `VOPD_Component`. This multiclass itself is a submulticlass, so `CheckTemplateArgValues` will be invoked on it.

`vOPDName` is passed as an empty string into here from `multiclass VOP2eInst`, and the typechecking caught this.

jayfoad replied:
> since fields are also resolved

That's the part I don't understand. It seems like CheckTemplateArgValues must do more than just checking types.

jayfoad replied:
Actually it seems like it is changing some behaviour. If I dump all the records from `AMDGPU.td` (_including_ your change to `VOPInstructions.td`) before and after your `TGParser.cpp` change, there are differences.

jofrn replied:
It is in the dynamic nature of TableGen. Typechecking resolves fields when it needs to do so. Adding CheckTemplateArgValues to this location mirrors the location it is at within `ParseSubClassReference`, so fields are also resolved when parsing non-multi sub classes.

jayfoad commented:
I think there is something weirder going on here, at least in the AMDGPU case. Consider:
```
class c {
  int cc = 0;
}
class d {
  int dd = 0;
}
multiclass m {
  def NAME : d;
}
defm a : m, c; // works!!!
defm b : c, m; // fails with error: Couldn't find multiclass 'c'
```
Is it ever supposed to be valid to use `defm` with a _class_ name, like `c` in `defm a : m, c;` above? Because that's exactly what we're doing here: https://github.com/llvm/llvm-project/blob/dbe47c2a06e0928edde802d062ecf1a0ce45fbb9/llvm/lib/Target/AMDGPU/VOP2Instructions.td#L153
```
multiclass
    VOP2Inst_e32_VOPD<string opName, VOPProfile P, bits<5> VOPDOp,
                      string VOPDName, SDPatternOperator node = null_frag,
                      string revOp = opName> {
  defm NAME : VOP2Inst_e32<opName, P, node, revOp>,
              VOPD_Component<VOPDOp, VOPDName>;
}
```
where `VOPD_Component` is a class.

jayfoad commented:
> Is it ever supposed to be valid to use defm with a class name, like c in defm a : m, c; above?

Oh, I see this is allowed and documented. Sorry for the noise.

arsenm commented on llvm/test/TableGen/submulticlass-typecheck.td:
```diff
@@ -0,0 +1,12 @@
+// RUN: not llvm-tblgen %s 2>&1 | FileCheck %s
+// XFAIL: vg_leak
+// CHECK: {{.*}}: error: Value specified for template argument 'B::op' is of type bits<4>; expected type bits<8>: C::op
```
Can you check the line number in the message

jofrn committed:
Give _fake16 record a vOPDName so that it now has a suffix in VOPDName.

The record now looks like:

  def V_CNDMASK_B16_e32 {
    ...
    Instruction BaseVOP = V_CNDMASK_B16_e32;
    string VOPDName = "v_dual_";
    bits<5> VOPDOp = { 1, 1, 1, 1, 1 };
    bit CanBeVOPDX = 0;
    -->
    Instruction BaseVOP = V_CNDMASK_B16_e32;
    string VOPDName = "v_dual_cndmask_b16";
    bits<5> VOPDOp = { 1, 1, 1, 1, 1 };
    bit CanBeVOPDX = 0;
  }

Whereas before, these fields were omitted entirely.

jofrn committed:
Add line number to assertion

jofrn commented on llvm/lib/Target/AMDGPU/VOP2Instructions.td:
```diff
@@ -319,7 +319,7 @@ multiclass
     VOP2eInst<string opName, VOPProfile P, SDPatternOperator node = null_frag,
               string revOp = opName, bit useSGPRInput = !eq(P.NumSrcArgs, 3)>
-    : VOP2eInst_Base<opName, P, -1, "", node, revOp, useSGPRInput>;
+    : VOP2eInst_Base<opName, P, -1, "v_cndmask_b16", node, revOp, useSGPRInput>;
```
We can pass vOPDName explicitly here in order to give this "fake" record more information, `defm V_CNDMASK_B16 : VOP2eInst <"v_cndmask_b16", VOP2e_I16_I16_I16_I1_fake16>;`. I do not see it as necessary though. The opcode is used in one location and it would not have been able to query this information before this change anyway. `VOPDOp` is given a value of `-1`, indicating a fake record as well.

arsenm replied:
I don't follow why this is hardcoding v_cndmask_b16

jofrn replied:
Otherwise the empty string will segfault on `!substr`. We can do this or check the indices, and the class is only used in one location.

arsenm replied:
substr segfaults on an empty string!? That's a really bad tablegen bug

jofrn replied:
Yes, replaced it with `v_` instead. Thanks.

jayfoad replied:
> substr segfaults on an empty string!? That's a really bad tablegen bug.

@jofrn if you have time can fix this too please (in a separate patch).

jofrn replied:
Thanks. I merged at this point and fixed the `-1` -> `0` switch you made at the end.

jofrn committed:
Change vOPDName to be prefix that is ignored by !substr in fake record.

Now the record has this field as before, and the fake part is included
in the record:

  string VOPDName = "v_dual_";

jayfoad commented:
The AMDGPU changes highlight a change in behaviour caused by your patch. Here's a simple example:
```
class C {
  bit x;
}
multiclass M1<bits<8> Val> {
  let x = !eq(Val, -1) in def NAME : C;
}
multiclass M2 : M1<-1>;
defm X : M2;
```
Before your patch I get:
```
def X {	// C
  bit x = 1;
}
```
After your patch I get:
```
def X {	// C
  bit x = 0;
}
```
Why did this change? Is the new behaviour more correct than the old behaviour? Can you please also add something like this as a test case?

In any case the AMDGPU code should probably be changed not to rely on equality of a `bits<5>` value and `-1`.

jofrn committed:
Add let !eq in submulticlass test.

jofrn commented:
> 
> Before your patch I get:
> 
> ```
> def X {	// C
>   bit x = 1;
> }
> ```
I not am getting `bit x = 1` in the aforementioned case, I get `bit x = 0` as is now both before and after. Uploading a test to show that one needs a `!cast` to perform the correct check.

jofrn commented on llvm/test/TableGen/submulticlass-leteq.td:
```diff
@@ -0,0 +1,21 @@
+// RUN: llvm-tblgen %s | FileCheck %s
+// XFAIL: vg_leak
+// CHECK:      def X0 { // C
+// CHECK-NEXT:   bit x = 1;
```
This is `bit x = 0;` in the before case, yet `!cast<bits<8>>(-1)` is identical to `Val`.

jofrn commented on llvm/test/TableGen/submulticlass-leteq.td:
```diff
@@ -0,0 +1,21 @@
+  let x = !eq(Val, !cast<bits<8>>(-1)) in def NAME : C;
+}
+multiclass M1<bits<8> Val> {
+  let x = !eq(Val, -1) in def NAME : C;
```
This should evaluate to `bit x = 0` because `-1` is and integer and `Val` is `255`.

jayfoad commented:
> The AMDGPU fix still doesn't look right. The intention of the code is clearly that when you pass in `-1`, the `!eq(..., -1)` test should be `true`.

How about #113279 to completely avoid this AMDGPU problem?

jofrn committed:
Remove prefix in fake record and rebase to b3acb25.

jayfoad commented on llvm/lib/Target/AMDGPU/VOPInstructions.td:
```diff
@@ -1931,4 +1931,4 @@ def VOPTrue16Table : GenericTable {
   let PrimaryKey = ["Opcode"];
   let PrimaryKeyName = "getTrue16OpcodeHelper";
-}
+}
```
Remove this change.

jayfoad reviewed (commented):
I haven't looked at the ARM parts but the rest LGTM.

jofrn committed:
Use the more derived type

jayfoad reviewed (approved):
LGTM - thanks and sorry for the very long delay!

jayfoad commented on llvm/lib/TableGen/TGParser.cpp:
Looks like this needs a tweak when you rebase:
```suggestion
  if (CheckTemplateArgValues(Result.TemplateArgs, ArgLocs,
```

jofrn commented:
Continued in https://github.com/llvm/llvm-project/pull/197128.

arsenm commented:
> #197128

Should this be closed then?
